arparser/src/parser.sk

```diff
@@ -0,0 +1,614 @@
+
+      this.!next_entry_start =
+        this.next_entry_start +
+        header_len +
+        header.size +
+        (header.size % 2);
+
+      if (this.is_name_table_id(header.identifier)) {
+        continue
+      };
+
+      entry = HeaderAndLocation{
+        header,
+        header_start,
+        data_start => header_start + header_len,
+      };
+
+      if (this.is_symbol_lookup_table_id(header.identifier)) {
+        this.!symbol_table_header = Some(entry)
+      } else {
+        this.headers.push(entry)
+      }
+    };
+
+    this.!scanned = true;
+    Success(void)
+  }
+
+  private mutable fun parse_gnu_extended_identifier(
+    header: Header,
+  ): Result<String, Error> {
+    start = parse_number(
+      header.identifier
+        .bytes()
+        .slice(kENTRY_IDENTIFIER_OFFSET + 1, kENTRY_IDENTIFIER_LEN)
+        .collect(Array),
+      10,
+    ) match {
+    | Failure(err) -> return Failure(err)
+    | Success(v) -> v
+    };
+
+    end = this.name_table.slice(start).findIndex(ch ->
+      ch == UInt8::truncate('/'.code())
+    ) match {
+    | Some(len) -> start + len
+    | None _ -> this.name_table.size()
+    };
+
+    // TODO: Explicit ASCII check.
+    identifier = String::fromUtf8(
+      this.name_table.slice(start, end).collect(Array),
+    );
+
+    Success(identifier)
+  }
+
+  private mutable fun parse_bsd_extended_identifier(
+    header: Header,
+  ): Result<(String, Int), Error> {
+    id_len = parse_number(
+      header.identifier
+        .bytes()
+        .slice(kENTRY_IDENTIFIER_OFFSET + 3, kENTRY_IDENTIFIER_LEN)
+        .collect(Array),
+      10,
+    ) match {
+    | Failure(err) -> return Failure(err)
+    | Success(v) -> v
+    };
+    if (header.size < id_len) {
+      return Failure(
+        ParseError(
+          `Entry size (${
+            header.size
+          }) smaller than extended entry identifier length (${id_len})`,
+        ),
+      )
+    };
+
+    id_buf = Array::mfill(id_len, UInt8::truncate(0));
+    this.reader.read_exact(id_buf.mbytes()) match {
+    | Failure(err) -> return Failure(IOError(err))
+    | Success _ -> void
+    };
+    // TODO: Explicit ASCII check.
+    identifier = String::fromUtf8(id_buf.collect(Array));
+
+    Success((identifier, id_len))
+  }
+
+  private readonly fun is_name_table_id(identifier: String): Bool {
+    this.variant is VariantGNU _ && identifier == kGNU_NAME_TABLE_ID
+  }
+
+  private readonly fun is_symbol_lookup_table_id(identifier: String): Bool {
+    this.variant match {
+    | VariantCommon() -> false
+    | VariantBSD() ->
+      identifier == kBSD_SYMBOL_LOOKUP_TABLE_ID ||
+        identifier == kBSD_SORTED_SYMBOL_LOOKUP_TABLE_ID
+    | VariantGNU() -> identifier == kGNU_SYMBOL_LOOKUP_TABLE_ID
+    }
+  }
+
+  private mutable fun parse_global_header(): Result<void, Error> {
+    this.parse_fixed_tag(kGLOBAL_HEADER) match {
+    | Failure(err) -> return Failure(err)
+    | Success _ -> void
+    };
+    this.!started = true;
```

Comment:
Similarly, I think `started` is useless because you start as you construct an `Archive`. Probably a remain from a past lazy parsing?
Also, you never use its value